Python/Product/PythonTools/PythonTools/Navigation/PythonLibraryManager.cs

```diff
@@ -41,42 +45,122 @@ internal interface IPythonLibraryManager : ILibraryManager {
-            if (analyzer.GetAnalysisEntryFromPath(task.FileName) == null) {
-                analyzer.AnalyzeFileAsync(task.FileName).ContinueWith(x => {
-                    item = x.Result;
-
-                    if (item != null) {
-                        // We subscribe to OnNewAnalysis here instead of OnNewParseTree so that 
-                        // in the future we can use the analysis to include type information in the
-                        // object browser (for example we could include base type information with
-                        // links elsewhere in the object browser).
-                        item.AnalysisComplete += (sender, args) => {
-                            _package.GetUIThread().InvokeAsync(() => FileParsed(task))
-                                .HandleAllExceptions(_package, GetType())
-                                .DoNotWait();
-                        };
+                    .GetPythonProject();
+            if (project == null) {
+                return;
+            }
+
+            AnalysisCompleteHandler handler;
+            lock (_handlers) {
+                if (!_handlers.TryGetValue(project, out handler)) {
+                    _handlers[project] = handler = new AnalysisCompleteHandler(this, project);
+                }
+            }
+
+            handler.AddTask(task);
+        }
+
+        sealed class AnalysisCompleteHandler : IDisposable {
+            public readonly PythonProjectNode Project;
+            private readonly PythonLibraryManager _owner;
+            private readonly Dictionary<string, LibraryTask> _tasks;
+            private VsProjectAnalyzer _analyzer;
+
+            public AnalysisCompleteHandler(PythonLibraryManager owner, PythonProjectNode project) {
+                Project = project ?? throw new ArgumentNullException(nameof(project));
+                _owner = owner ?? throw new ArgumentNullException(nameof(owner));
+                _tasks = new Dictionary<string, LibraryTask>(StringComparer.OrdinalIgnoreCase);
+                Project.ProjectAnalyzerChanging += Project_ProjectAnalyzerChanging;
+                _analyzer = Project.TryGetAnalyzer();
+                if (_analyzer != null) {
+                    _analyzer.AnalysisComplete += Analyzer_AnalysisComplete;
+                }
+            }
+
+            public void AddTask(LibraryTask task) {
+                lock (_tasks) {
+                    _tasks[task.FileName] = task;
+                }
+            }
+
+            public void Dispose() {
+                lock (this) {
+                    Project.ProjectAnalyzerChanging -= Project_ProjectAnalyzerChanging;
+                    if (_analyzer != null) {
+                        _analyzer.AnalysisComplete -= Analyzer_AnalysisComplete;
                     }
-                });
+                }
+            }
+
+            private void Project_ProjectAnalyzerChanging(object sender, AnalyzerChangingEventArgs e) {
+                lock (this) {
```

Comment:
>this [](start = 22, length = 4)

I know it's a private class, all usage of it is nearby and it's unlikely that anyone outside the class will try to lock on an AnalysisCompleteHandler object from outside, but I'm still not a fan of lock(this).

Reply:
Yeah, me neither, but it's an inner class instance so seemed better than allocating another new object just for locking the event handlers.